Use more smart pointers for members

rwlbuis opened the pull request:
#### d827195d613f685f3af939eeae5c779c27dee3ac
<pre>
Use more smart pointers for members
<a href="https://bugs.webkit.org/show_bug.cgi?id=274252">https://bugs.webkit.org/show_bug.cgi?id=274252</a>

Reviewed by Chris Dumez.

Use more smart pointers for members to address [clang-analyzer-webkit.NoUncountedMemberChecker] warnings.

* Source/WebCore/animation/KeyframeEffect.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::createParser):
* Source/WebCore/html/HTMLFrameOwnerElement.h:
(WebCore::SubframeLoadingDisabler::SubframeLoadingDisabler):
(WebCore::SubframeLoadingDisabler::~SubframeLoadingDisabler):
* Source/WebCore/html/HTMLProgressElement.cpp:
(WebCore::HTMLProgressElement::HTMLProgressElement):
* Source/WebCore/html/HTMLProgressElement.h:
* Source/WebCore/html/HTMLTrackElement.cpp:
(WebCore::HTMLTrackElement::~HTMLTrackElement):
(WebCore::HTMLTrackElement::track):
* Source/WebCore/html/HTMLTrackElement.h:
* Source/WebCore/html/track/LoadableTextTrack.cpp:
(WebCore::LoadableTextTrack::LoadableTextTrack):
(WebCore::LoadableTextTrack::trackElementIndex):
* Source/WebCore/html/track/LoadableTextTrack.h:
* Source/WebCore/inspector/DOMPatchSupport.cpp:
(WebCore::DOMPatchSupport::patchDocument):
* Source/WebCore/rendering/RenderView.h:
* Source/WebCore/rendering/updating/RenderTreeUpdater.cpp:
(WebCore::RenderTreeUpdater::commit):
(WebCore::RenderTreeUpdater::updateElementRenderer):
(WebCore::RenderTreeUpdater::createRenderer):
(WebCore::RenderTreeUpdater::createTextRenderer):
(WebCore::RenderTreeUpdater::updateRenderViewStyle):
(WebCore::RenderTreeUpdater::renderView):
* Source/WebCore/rendering/updating/RenderTreeUpdater.h:
* Source/WebCore/xml/parser/XMLDocumentParser.h:
* Source/WebCore/xml/parser/XMLDocumentParserLibxml2.cpp:
(WebCore::XMLDocumentParser::XMLDocumentParser):
(WebCore::XMLDocumentParser::endElementNs):

Canonical link: <a href="https://commits.webkit.org/281261@main">https://commits.webkit.org/281261@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/32c19b530b4c87b3a6fb2a688d610e27080139ea

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/59233 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/38576 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/11752 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/63138 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/9676 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/61362 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/46230 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/9906 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/48118 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/6901 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/61263 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/36054 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/51269 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/28940 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/32767 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/8528 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/8680 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/54725 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/8808 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/64910 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/3161 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/8749 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/55501 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/3172 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/51270 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/55589 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13134 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/2638 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/34392 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/35475 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/36561 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/35220 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/xml/parser/XMLDocumentParser.h:
```diff
@@ -153,7 +153,7 @@ class XMLDocumentParser final : public ScriptableDocumentParser, public PendingS
     xmlParserCtxtPtr context() const { return m_context ? m_context->context() : nullptr; };
 
-    LocalFrameView* m_view { nullptr };
+    RefPtr<LocalFrameView> m_view;
```
Slightly worried about a reference cycle here given that the XMLDocumentParser is not a stack-only object and is held strongly by the Document.

geoffreygaren replied:
Document holds WeakPtr to Frame, and Frame holds RefPtr to LocalFrameView. So I think RefPtr is probably fine. But WeakPtr is safer against cycles, and more consistent with Document holding WeakPtr to Frame. So let's make m_view WeakPtr here.

Or you could go even further and have a "bool m_isInFrameView" data member instead of a pointer. As far as I can tell, XMLDocumentParser literally only uses m_view to remember if it was initialized with a FrameView.

rwlbuis replied:
I like the "bool m_isInFrameView" idea, it is what I implemented, but if you prefer WeakPtr, that would work too.

cdumez commented on Source/WebCore/html/HTMLProgressElement.cpp:
```diff
@@ -42,7 +42,6 @@ const double HTMLProgressElement::InvalidPosition = -2;
 HTMLProgressElement::HTMLProgressElement(const QualifiedName& tagName, Document& document)
     : HTMLElement(tagName, document, TypeFlag::HasCustomStyleResolveCallbacks)
-    , m_value(0)
     , m_isDeterminate(false)
```
Seems unnecessary too.

rwlbuis replied:
Well spotted, removed.

cdumez commented on Source/WebCore/xml/parser/XMLDocumentParser.h:
```diff
@@ -66,9 +66,9 @@ class XMLDocumentParser final : public ScriptableDocumentParser, public PendingS
     WTF_OVERRIDE_DELETE_FOR_CHECKED_PTR(XMLDocumentParser);
 public:
-    static Ref<XMLDocumentParser> create(Document& document, LocalFrameView* view, OptionSet<ParserContentPolicy> policy = DefaultParserContentPolicy)
+    static Ref<XMLDocumentParser> create(Document& document, bool isInFrameView, OptionSet<ParserContentPolicy> policy = DefaultParserContentPolicy)
```
This goes against our usual style. We should not be passing boolean parameters like this. Either we keep the pointer or we switch to an enum class.

rwlbuis replied:
I went for the enum option.

rwlbuis committed:
Use more smart pointers for members
https://bugs.webkit.org/show_bug.cgi?id=274252

Reviewed by Chris Dumez.

Use more smart pointers for members to address [clang-analyzer-webkit.NoUncountedMemberChecker] warnings.

* Source/WebCore/animation/KeyframeEffect.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::createParser):
* Source/WebCore/html/HTMLFrameOwnerElement.h:
(WebCore::SubframeLoadingDisabler::SubframeLoadingDisabler):
(WebCore::SubframeLoadingDisabler::~SubframeLoadingDisabler):
* Source/WebCore/html/HTMLProgressElement.cpp:
(WebCore::HTMLProgressElement::HTMLProgressElement):
* Source/WebCore/html/HTMLProgressElement.h:
* Source/WebCore/html/HTMLTrackElement.cpp:
(WebCore::HTMLTrackElement::~HTMLTrackElement):
(WebCore::HTMLTrackElement::track):
* Source/WebCore/html/HTMLTrackElement.h:
* Source/WebCore/html/track/LoadableTextTrack.cpp:
(WebCore::LoadableTextTrack::LoadableTextTrack):
(WebCore::LoadableTextTrack::trackElementIndex):
* Source/WebCore/html/track/LoadableTextTrack.h:
* Source/WebCore/inspector/DOMPatchSupport.cpp:
(WebCore::DOMPatchSupport::patchDocument):
* Source/WebCore/rendering/RenderView.h:
* Source/WebCore/rendering/updating/RenderTreeUpdater.cpp:
(WebCore::RenderTreeUpdater::commit):
(WebCore::RenderTreeUpdater::updateElementRenderer):
(WebCore::RenderTreeUpdater::createRenderer):
(WebCore::RenderTreeUpdater::createTextRenderer):
(WebCore::RenderTreeUpdater::updateRenderViewStyle):
(WebCore::RenderTreeUpdater::renderView):
* Source/WebCore/rendering/updating/RenderTreeUpdater.h:
* Source/WebCore/xml/parser/XMLDocumentParser.h:
* Source/WebCore/xml/parser/XMLDocumentParserLibxml2.cpp:
(WebCore::XMLDocumentParser::XMLDocumentParser):
(WebCore::XMLDocumentParser::endElementNs):

Canonical link: https://commits.webkit.org/281261@main
